File: app/src/main/java/org/zeroxlab/momodict/archive/TarExtractor.java
package org.zeroxlab.momodict.archive;

import org.apache.commons.compress.archivers.tar.TarArchiveEntry;
import org.apache.commons.compress.archivers.tar.TarArchiveInputStream;
import org.apache.commons.compress.compressors.gzip.GzipCompressorInputStream;
import org.apache.commons.io.FileUtils;

import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;

import rx.functions.Action0;

public class TarExtractor implements Extractor {

    @Override
    public FileSet extract(final File outputDir, InputStream inputStream) throws Exception {
        FileSet archive = new FileSet();
        archive.setCleanCallback(new Action0() {
            @Override
            public void call() {
                try {
                    if (outputDir.isDirectory()) {
                        FileUtils.deleteDirectory(outputDir);
                    }
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        });

        TarArchiveInputStream tis = new TarArchiveInputStream(inputStream);
        while (tis.getNextEntry() != null) {
            onEntryFound(outputDir, tis, tis.getCurrentEntry(), archive);
        }

        if (!archive.isSane()) {
            throw new Exception("RealmBook is malformed");
        }
        return archive;
    }

    private void onEntryFound(File parent,
                              InputStream is,
                              TarArchiveEntry entry,
                              FileSet archive)
            throws Exception {

        final String fileName = entry.getName();
        if (entry.isDirectory()) {
            File dir = new File(parent, fileName);
            if (!dir.mkdirs()) {
                throw new Exception("Create dir fail");
            }
        } else {
            // write file
            File out = new File(parent, fileName);

            byte[] buf = new byte[65535];
            FileOutputStream fos = new FileOutputStream(out);
            BufferedOutputStream bos = new BufferedOutputStream(fos);
            int count;
            while ((count = is.read(buf)) != -1) {
                bos.write(buf, 0, count);
            }
            bos.close();

            // cache file path
            if (fileName.endsWith(".idx")) {
                archive.set(FileSet.Type.IDX, out.getAbsolutePath());
            } else if (fileName.endsWith(".ifo")) {
                archive.set(FileSet.Type.IFO, out.getAbsolutePath());
            } else if (fileName.endsWith(".dict.dz")) {
                // found a dz file, use gzip to extract again
                String extractFileName = out.getAbsolutePath().replace(".dict.dz", ".dict");
                FileInputStream fis = new FileInputStream(out);
                GzipCompressorInputStream gis = new GzipCompressorInputStream(fis);
                FileOutputStream fos2 = new FileOutputStream(extractFileName);
                byte[] buffer = new byte[2048];
                int r = 0;
                while ((r = gis.read(buffer)) != -1) {
                    fos2.write(buffer, 0, r);
                }
                fis.close();
                fos2.close();
                archive.set(FileSet.Type.DICT, extractFileName);
            } else if (fileName.endsWith(".dict")) {
                archive.set(FileSet.Type.DICT, out.getAbsolutePath());
            }
        }
        System.out.println(entry.getName());
    }
}

File: app/src/main/java/org/zeroxlab/momodict/archive/TarExtractor.kt
package org.zeroxlab.momodict.archive

import org.apache.commons.compress.archivers.tar.TarArchiveEntry
import org.apache.commons.compress.archivers.tar.TarArchiveInputStream
import org.apache.commons.compress.compressors.gzip.GzipCompressorInputStream
import org.apache.commons.io.FileUtils
import rx.functions.Action0
import java.io.BufferedOutputStream
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.io.InputStream

class TarExtractor : Extractor {

    @Throws(Exception::class)
    override fun extract(outputDir: File, inputStream: InputStream): FileSet {
        val archive = FileSet()
        archive.setCleanCallback(Action0 {
            try {
                if (outputDir.isDirectory) {
                    FileUtils.deleteDirectory(outputDir)
                }
            } catch (e: IOException) {
                e.printStackTrace()
            }
        })

        val tis = TarArchiveInputStream(inputStream)
        while (tis.nextEntry != null) {
            onEntryFound(outputDir, tis, tis.currentEntry, archive)
        }

        if (!archive.isSane) {
            throw Exception("Book is malformed")
        }
        return archive
    }

    @Throws(Exception::class)
    private fun onEntryFound(parent: File,
                             iStream: InputStream,
                             entry: TarArchiveEntry,
                             archive: FileSet) {

        val fileName = entry.name
        if (entry.isDirectory) {
            val dir = File(parent, fileName)
            if (!dir.mkdirs()) {
                throw Exception("Create dir fail")
            }
        } else {
            // write file
            val out = File(parent, fileName)

            val buf = ByteArray(65535)
            val fos = FileOutputStream(out)
            val bos = BufferedOutputStream(fos)
            var count: Int = iStream.read(buf)
            while (count != -1) {
                bos.write(buf, 0, count)
                count = iStream.read(buf)
            }
            bos.close()

            // cache file path
            if (fileName.endsWith(".idx")) {
                archive[FileSet.Type.IDX] = out.absolutePath
            } else if (fileName.endsWith(".ifo")) {
                archive[FileSet.Type.IFO] = out.absolutePath
            } else if (fileName.endsWith(".dict.dz")) {
                // found a dz file, use gzip to extract again
                val extractFileName = out.absolutePath.replace(".dict.dz", ".dict")
                val fis = FileInputStream(out)
                val gis = GzipCompressorInputStream(fis)
                val fos2 = FileOutputStream(extractFileName)
                val buffer = ByteArray(2048)
                var r = gis.read(buffer)
                while (r != -1) {
                    fos2.write(buffer, 0, r)
                    r = gis.read(buffer)
                }
                fis.close()
                fos2.close()
                archive[FileSet.Type.DICT] = extractFileName
            } else if (fileName.endsWith(".dict")) {
                archive[FileSet.Type.DICT] = out.absolutePath
            }
        }
        println(entry.name)
    }
}
